### ISSUE TITLE
DeaDBeeF 0.7 resets user settings of geometry of widgets

### ISSUE THREAD

**Issue by KernelModule** [Alexey-Yakovenko/deadbeef#1465]
This is how it should be: [config_user.txt](https://github.com/Alexey-Yakovenko/deadbeef/files/112604/config_user.txt)
This is what you get after restart player [config.txt](https://github.com/Alexey-Yakovenko/deadbeef/files/112624/config.txt)

In screenshots:
User settings (approximately):
![before](https://cloud.githubusercontent.com/assets/9074462/12724836/d64508b0-c920-11e5-804e-0391eb8f1c45.png)

After restart:
![after](https://cloud.githubusercontent.com/assets/9074462/12724837/d6513572-c920-11e5-8fbe-996583a7f8a2.png)

**Comment by SydMax** [comment]
I have seen this problem for quite some time on the master branch builds.
Can you confirm certain detail in the bug - in the messed-up layout try to _expand_ lyrics widget even more to the left. In my experience that even more expanded layout is correctly preserved.
Also important detail:  config file gets corrupted during player shutdown, not in the start. So when you change layout I can see it being properly written down to the config file, but in the moment of shutting down values in config become corrupted.

**Comment by Alexey-Yakovenko** [comment]
I think the reason why this problem remained undetected by the devs is because it's caused by a rarely used 3rd party plugin, which writes something wrong in the config. But we'll find out soon.

**Comment by Alexey-Yakovenko** [comment]
@KernelModule just out of curiosity, if you replace the lyrics plugin with any other (standard) widget, does the problem disappear?

**Comment by SydMax** [comment]
I am using Infobar plugin, for example.
I also tried now to substitute corrupted config with preserved good one before sturting the player. Result - messed up layout still.
[config.good.txt](https://github.com/Alexey-Yakovenko/deadbeef/files/114211/config.good.txt)
[config.bad.txt](https://github.com/Alexey-Yakovenko/deadbeef/files/114210/config.bad.txt) - config after shutdown
[config.bad.restart.txt](https://github.com/Alexey-Yakovenko/deadbeef/files/114209/config.bad.restart.txt) - config of player started with config.good

**Comment by KernelModule** [comment]
@Alexey-Yakovenko 

> just out of curiosity, if you replace the lyrics plugin with any other (standard) widget, does the problem disappear?
No, bug still happens.

**Comment by Alexey-Yakovenko** [comment]
what about switching GTK2 -> 3, or the other way round?

**Comment by KernelModule** [comment]
@SydMax 
> Can you confirm certain detail in the bug - in the messed-up layout try to expand lyrics widget even more to the left. In my experience that even more expanded layout is correctly preserved.

Yes.

> Also important detail: config file gets corrupted during player shutdown, not in the start. So when you change layout I can see it being properly written down to the config file, but in the moment of shutting down values in config become corrupted.

Yes.

**Comment by KernelModule** [comment]
@Alexey-Yakovenko 
> what about switching GTK2 -> 3, or the other way round?

No changes.

**Comment by cboxdoerfer** [comment]
@Alexey-Yakovenko, I can reproduce that. And probably know the reason for that bug. The config both have maximised mode enabled, but apparently somehow during the initialization the layout gets created with the values of mainwin.geometry.w/h instead of a maximized window. And gtk shrinks gtk paned according to the window size, therefore the whole layout gets destroyed at restart.

**Comment by Alexey-Yakovenko** [comment]
I think the bug #975 is a duplicate of this.

**Comment by Alexey-Yakovenko** [comment]
@cboxdoerfer but they claim the config gets wrong values on quit, not on startup.. wrong?

**Comment by cboxdoerfer** [comment]
not sure if that's a bug on our side or maybe the window managers are to blame. I guess https://github.com/Alexey-Yakovenko/deadbeef/issues/1464 is also related to that.

**Comment by Alexey-Yakovenko** [comment]
There was definitely another duplicate, but can't find it. Maybe it's closed.

**Comment by cboxdoerfer** [comment]
Yes, maybe this bug gets triggered on shutdown (and maybe start too)

**Comment by KernelModule** [comment]
> I think the bug #975 is a duplicate of this.
I can't reproduce this bug in 0.6.

**Comment by Alexey-Yakovenko** [comment]
Perhaps the #1378 and #1432 are related.

**Comment by SydMax** [comment]
Aha!
Found the way to reproduce bug:
1. Make the player window rather small, ie half of the screen width
2. Maximize it. Now make the playlist widget to occupy more then half a screen width, so the lyrics widget will not fit into the window when it's not maximized.
 Now you can trigger the bug with shutdown, or just switch to the non-maximized window - it will cause it too

**Comment by Alexey-Yakovenko** [comment]
I just tried, and I can repro this bug very easily on the first try

**Comment by Alexey-Yakovenko** [comment]
The splitter value is writter to the config on quit, and it's already wrong, as stated by @SydMax

**Comment by SydMax** [comment]
Turns out, as I wrote, it is not even related to quit. Just un-maximizing

**Comment by Alexey-Yakovenko** [comment]
OH. I just noticed it's about the maximized window.. that explains why I didn't notice.
I never run deadbeef maximized. And the problem disappears as soon as I un-maximize it.

**Comment by SydMax** [comment]
Playing with the size of non-maximized window: basically when i am shrinking the window, when the right edge of the window runs into the splitter in the layout, splitter gets moved permanently, so increasing window size does not restore layout.

**Comment by cboxdoerfer** [comment]
@Alexey-Yakovenko, just checked this is what happens:

- have a running session of deadbeef maximised
- setup a layout
- quit player: the layout gets saved correctly
- start player: the layout config gets loaded correctly
- but still, the layout is wrong

So somewhere after we load the layout correctly the window gets resized and therefore the layout gets destroyed.

**Comment by Alexey-Yakovenko** [comment]
yes, but this is expected. maybe not what you expect -- but it's supposed to work this way.
so let's focus on the maximization issue for now.

**Comment by Alexey-Yakovenko** [comment]
@cboxdoerfer 

> quit player: the layout gets saved correctly

This is wrong. Layout doesn't get saved correctly. the splitter "pos" attribute is wrong at this time.

**Comment by cboxdoerfer** [comment]
I just tried it a second time, with debug output in the save code and looking at the config file: after quit the config is correct.

**Comment by Alexey-Yakovenko** [comment]
I think what happens is this:

just before quitting, the window gets hidden / unmaximized, which leads to the window size smaller than the splitter pos in maximized mode

The splitter shrinks

Then serialization happens -- and we get the wrong value.

This is what happens on my computer. It looks like you're getting something else @cboxdoerfer

**Comment by cboxdoerfer** [comment]
maybe the behaviour depends on the window manager

**Comment by cboxdoerfer** [comment]
I'll try it with a few

**Comment by SydMax** [comment]
I am using KWin, btw

**Comment by KernelModule** [comment]
> I am using KWin, btw
Same.

**Comment by Alexey-Yakovenko** [comment]
I added the fix which moves w_save call to before GTK deinitialization happens, separately from the w_free call. It works for me now.

**Comment by cboxdoerfer** [comment]
still broken on my machine

**Comment by SydMax** [comment]
But maximizing/un-maximizing still causes layout to change if some splitter does not fit into the unmaximized window, right?

**Comment by Alexey-Yakovenko** [comment]
@cboxdoerfer did you build off 0.7 branch?

**Comment by Alexey-Yakovenko** [comment]
> But maximizing/un-maximizing still causes layout to change if some splitter does not fit into the unmaximized window, right?

Yeah, this is a separate issue, which is basically bug #1432

**Comment by cboxdoerfer** [comment]
yes

**Comment by Alexey-Yakovenko** [comment]
Then you def have another issue, not this bug.

**Comment by Alexey-Yakovenko** [comment]
but let's wait for the OP to test it.

**Comment by cboxdoerfer** [comment]
well, I didn't get the config "corruption" on exit in the first place. So yeah, maybe that's something else here

**Comment by SydMax** [comment]
I'm not the OP, but my case and setup seems to be almost identical.
I have done fresh build of 0.7 branch. Oddly enough, problem is not resolved with proposed patch and still there: layout is changed at Quit from maximized state.

**Comment by SydMax** [comment]
@Alexey-Yakovenko sorry, but have to ask, are you shure that you tested proposed patch on the combination of layout and window size that is susceptible to corruption at unmaximize (bug #1432)

**Comment by Alexey-Yakovenko** [comment]
@SydMax which patch?

**Comment by Alexey-Yakovenko** [comment]
Hey @KernelModule have you tried the fix?
The build is here http://sourceforge.net/projects/deadbeef/files/travis/linux/0.7/deadbeef-static_0.7.1-beta1-1_x86_64.tar.bz2/download

**Comment by SydMax** [comment]
I meant this commit https://github.com/Alexey-Yakovenko/deadbeef/commit/a5da47d0ade1114527613ae8e2cf3c3e77e15e00

**Comment by XRevan86** [comment]
Apparently issue is caused by DeaDBeeF working with geometry of an unmaximised window even when maximised.
Size of unmaximised window in @KernelModule's config is small, much smaller than size his layout is made for, fixing that in his config fixes the issue.

**Comment by Alexey-Yakovenko** [comment]
@XRevan86 this is what the commit a5da47d fixes

**Comment by Alexey-Yakovenko** [comment]
@SydMax which specific problem are _you_ trying to solve? are you sure that it's the same as this bug report? After applying the fix, I can't reproduce it anymore.

**Comment by KernelModule** [comment]
> Hey @KernelModule have you tried the fix?
The build is here http://sourceforge.net/projects/deadbeef/files/travis/linux/0.7/deadbeef-static_0.7.1-beta1-1_x86_64.tar.bz2/download

Yes, and fix doesn't work for me. Also, launch deadbeef with proper config (which was copied while deadbeef is running with proper interface) seems also doesn't work, checksums doesn't changed.

**Comment by Alexey-Yakovenko** [comment]
thanks, will look further.

**Comment by SydMax** [comment]
I am trying to solve this problem:
layout is changed at Quit from maximized state _IF_ we have a combination of layout and unmaximized window size that is susceptible to corruption at unmaximize (ie layout is too big for unmaximized window)

**Comment by Alexey-Yakovenko** [comment]
Yes, I figured that out from the beginning. It's weird that my fix didn't work for you.

May I ask how exactly do you quit deadbeef? ctrl+q, or clicking the "X", or menu, etc?

**Comment by SydMax** [comment]
For me it doesn't matter. ctrl+q, and menu both cause the problem

**Comment by KernelModule** [comment]
> Yes, I figured that out from the beginning. It's weird that my fix didn't work for you.
May I ask how exactly do you quit deadbeef? ctrl+q, or clicking the "X", or menu, etc?

ctrl+q

**Comment by Alexey-Yakovenko** [comment]
gtk_quit is called multiple times.. that's the real reason.

**Comment by Alexey-Yakovenko** [comment]
(or not)

**Comment by KernelModule** [comment]
> gtk_quit is called multiple times.. that's the real reason.

If the reason only in quit, then why deadbeef with proper config at time of launch launches with bug?

**Comment by Alexey-Yakovenko** [comment]
no, it seems it's both on quit and on start.
I fixed the quit, but didn't fix the start.
Trying another fix now.

**Comment by Alexey-Yakovenko** [comment]
updated the fix, please try

**Comment by KernelModule** [comment]
> updated the fix, please try

Link for static build, thanks.

**Comment by Alexey-Yakovenko** [comment]
Still compiling, will post as soon as it's done (the URL is the same actually)

**Comment by SydMax** [comment]
Last fix resolved the issue for me.

**Comment by Alexey-Yakovenko** [comment]
@KernelModule the build is done http://sourceforge.net/projects/deadbeef/files/travis/linux/0.7/deadbeef-static_0.7.1-beta1-1_x86_64.tar.bz2/download

**Comment by SydMax** [comment]
The same problem with layout corruption was triggered also at closing/opening into tray (xembed tray icon) from the maximized window. Is it possible to add the same workaround there?

**Comment by Alexey-Yakovenko** [comment]
It's more complicated with the tray.

The problem is that the window really gets un-maximized / restored, when going out of hidden state, which causes splitter to be moved to fit in the window geometry.

We can try reloading the widget layout after "restore from tray" is called, but this is not something that's supposed to work multiple times.

It's supposed to be loaded only once, as it actually creates the widgets.

**Comment by Alexey-Yakovenko** [comment]
I created the new bug for fixing systray issue

**Comment by KernelModule** [comment]
@Alexey-Yakovenko Thanks, last fix is work for me. But issue stiil exist with closing/opening into tray, just as said @SydMax.

**Comment by Alexey-Yakovenko** [comment]
ok, closing the issue, and let's move the tray icon discussion to the new bug (linked above)
